src/source/schematron.c

```diff
@@ -100,8 +107,342 @@ struct oscap_schema_table_entry OSCAP_SCHEMATRON_TABLE[] = {
+		if (!oscap_str_startswith(xlink_href, "#"))
+			res = true;
+		free(xlink_href);
+	}
+	return res;
+}
+
+static bool _req_src_236_2_sub6(xmlNodePtr component, xmlXPathContextPtr context, const char *name, const char *rule_id)
+{
+	bool exists_ocil = false;
+	char *ocil_xpath = oscap_sprintf(".//ocil:questionnaire[@id='%s']", name);
+	xmlXPathObjectPtr ocil_questionnaires = xmlXPathNodeEval(component, BAD_CAST ocil_xpath, context);
+	for (int m = 0; m < ocil_questionnaires->nodesetval->nodeNr; m++) {
+		xmlNodePtr ocil_questionnaire = ocil_questionnaires->nodesetval->nodeTab[m];
+		char *ocil_questionnaire_id = (char *) xmlGetProp(ocil_questionnaire, BAD_CAST "id");
+		dD("Found OCIL questionnaire id='%s' for rule id='%s'", ocil_questionnaire_id, rule_id);
+		free(ocil_questionnaire_id);
+		exists_ocil = true;
+	}
+	free(ocil_xpath);
+	xmlXPathFreeObject(ocil_questionnaires);
+	return exists_ocil;
+}
+
+static bool _req_src_236_2_sub5(xmlNodePtr component, xmlXPathContextPtr context, const char *name, const char *rule_id)
+{
+	/* exists(xcf:get-component(xcf:get-component-ref($m/cat:catalog, $o/@href))//oval-def:definition[@id eq $o/@name and matches(@class, '^(compliance|patch)$')]) */
+	bool exists_oval = false;
+	char *oval_xpath = oscap_sprintf(".//oval-def:definition[@id='%s' and (@class='compliance' or @class='patch')]", name);
+	xmlXPathObjectPtr oval_definitions = xmlXPathNodeEval(component, BAD_CAST oval_xpath, context);
+	for (int m = 0; m < oval_definitions->nodesetval->nodeNr; m++) {
+		xmlNodePtr oval_definition = oval_definitions->nodesetval->nodeTab[m];
+		char *oval_def_id = (char *) xmlGetProp(oval_definition, BAD_CAST "id");
+		dD("Found OVAL definition id='%s' for rule id='%s'", oval_def_id, rule_id);
+		free(oval_def_id);
+		exists_oval = true;
+	}
+	free(oval_xpath);
+	xmlXPathFreeObject(oval_definitions);
+	return exists_oval;
+}
+
+static bool _req_src_236_2_sub4(xmlNodePtr check_content_ref_node, xmlNodePtr catalog, xmlXPathContextPtr context, const char *rule_id)
+{
+	char *href = (char *) xmlGetProp(check_content_ref_node, BAD_CAST "href");
+	/* xcf:get-component(xcf:get-component-ref($m/cat:catalog, $o/@href) */
+	xmlNodePtr component_ref =  _xcf_get_component_ref(catalog, href, context);
+	free(href);
+	if (component_ref == NULL) {
+		dD("component_ref not found");
+		return false;
+	}
+	xmlNodePtr component = _xcf_get_component(component_ref, context);
+	if (component == NULL) {
+		dD("component not found\n");
+		return false;
+	}
+
+	char *name = (char *) xmlGetProp(check_content_ref_node, BAD_CAST "name");
+
+	/* exists(xcf:get-component(xcf:get-component-ref($m/cat:catalog, $o/@href))//oval-def:definition[@id eq $o/@name and matches(@class, '^(compliance|patch)$')]) or exists(xcf:get-component(xcf:get-component-ref($m/cat:catalog, $o/@href))//ocil:questionnaire[@id eq $o/@name])) */
+	bool res = (_req_src_236_2_sub5(component, context, name, rule_id) || _req_src_236_2_sub6(component, context, name, rule_id));
```

Comment:
I believe you should test which kind the retrieved component is.

For example, if you try to run this xpath query `".//ocil:questionnaire[@id='%s']"` using an OVAL component it yields:

`XPath error : Undefined namespace prefix`

The datastream: `R1300/r1300.2.2-datastream.xml.result.xml` line 108 contains the href:

`<xccdf:check-content-ref href="495dA_374l7P7640Z8oZuU27CQp049194N6-0585Ob1621ui41540dv2Dt"`, if you remove the last char `t` it will correctly find the component (OVAL one) and the fail does not occur in the line schematron.c:283 anymore. I don't know if this typo in the datastream is intended, but it contains the OVAL definition name that is missing from the OVAL component, which is the core assert of SRC-346-1 (to check if a OVAL definition exists in the OVAL component)

Reply:
I think the problem was that I haven't registered the `ocil:` namespace prefix. I have fixed  that below.

> I believe you should test which kind the retrieved component is.

I think that if `_req_src_236_2_sub5` or `_req_src_236_2_sub6` don't find the element they will return false. And their calls are a part of a disjunction.

Reply:
That seems to fix the problem. But do you think that the potential typo in the `R1300/r1300.2.2-datastream.xml.result.xml` is intentional or not? with that character in place, the datastream is unable to be processed and find the correct OVAL component. Nevertheless, the implementation seems to be working in detecting if the oval definition is present or not.

Reply:
I think that's intentional there to check if the tested product is able to detect this.

Reply:
Okay, it sort of makes sense.